### PR TITLE
[REVIEW] introduce kafka consumer and meta client to custreamz

### LINKED ISSUE #5769: [FEA] Create librdkafka wrapper in custreamz
Now that the accelerated kafka reader is merged into cudf a wrapper needs to be placed in the custreamz python project that acts as a shim for python streamz making it possible for that logic to be seamlessly used with python streamz itself as a none default engine.

This "shim code" will need to follow the exact paradigms and naming conventions that Python streamz currently already uses for kafka to ensure backwards compatability.

### PR DESCRIPTION
**Pull request by jdye64**
Add support for creating cudf dataframes directly from kafka. This includes wrapping the existing `kafka_consumer` datasource in cudf for reading the dataframe and also adding a meta client connection to handle operations unrelated to cudf like committing and receiving kafka offsets to manage the streaming data flow.

This closes #5769

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: python/custreamz/custreamz/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+from .kafka import Consumer
```
Comment (kkraus14) [inline python/custreamz/custreamz/__init__.py:1]: Add copyright

File: python/custreamz/custreamz/kafka.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    def __init__(self, kafka_configs):
+        self.kafka_configs = kafka_configs
+
+        self.kafka_confs = {}
+        for key, value in self.kafka_configs.items():
+            self.kafka_confs[str.encode(key)] = str.encode(value)
+
+        self.kafka_meta_client = KafkaDatasource(self.kafka_confs)
+
+    @docutils.doc_unsubscribe()
```
Comment (kkraus14) [inline python/custreamz/custreamz/kafka.py:21]: Any reason you're putting these in a separate file? Would be better to just inline them here unless there's a reason not to.

File: python/custreamz/custreamz/kafka.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+
+# Base class for anything class that needs to interact with Apache Kafka
+class CudfKafkaClient(object):
+    def __init__(self, kafka_configs):
+        self.kafka_configs = kafka_configs
+
+        self.kafka_confs = {}
+        for key, value in self.kafka_configs.items():
+            self.kafka_confs[str.encode(key)] = str.encode(value)
```
Comment (kkraus14) [inline python/custreamz/custreamz/kafka.py:17]: Can use a dictionary comprehension here instead:
```suggestion
        self.kafka_confs = {str.encode(key): str.encode(value) for key, value in self.kafka_configs.items()}
```

File: python/custreamz/custreamz/kafka.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+from cudf_kafka._lib.kafka import KafkaDatasource
+
+import cudf
+
+from custreamz.utils import docutils
+
+
+# Base class for anything class that needs to interact with Apache Kafka
+class CudfKafkaClient(object):
+    def __init__(self, kafka_configs):
```
Comment (kkraus14) [inline python/custreamz/custreamz/kafka.py:12]: Can you add a docstring to the `__init__` method here describing the class and its initialization?

File: python/custreamz/custreamz/kafka.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+import confluent_kafka as ck
+from cudf_kafka._lib.kafka import KafkaDatasource
+
+import cudf
+
+from custreamz.utils import docutils
+
+
+# Base class for anything class that needs to interact with Apache Kafka
+class CudfKafkaClient(object):
```
Comment (kkraus14) [inline python/custreamz/custreamz/kafka.py:11]: Don't need to subclass `object` as of Python 3:
```suggestion
class CudfKafkaClient:
```

File: python/custreamz/custreamz/kafka.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    def close(self, timeout=10000):
+
+        """{docstring}"""
+
+        return self.kafka_meta_client.close(timeout=timeout)
+
+
+# Apache Kafka Consumer implementation
+class Consumer(CudfKafkaClient):
+    def __init__(self, kafka_configs):
```
Comment (kkraus14) [inline python/custreamz/custreamz/kafka.py:38]: Add docstring

File: python/custreamz/custreamz/kafka.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        if message_format.lower() == "json":
+            result = cudf.io.read_json(self.kafka_datasource)
+        elif message_format.lower() == "csv":
+            result = cudf.io.read_csv(self.kafka_datasource)
+        elif message_format.lower() == "orc":
+            result = cudf.io.read_orc(self.kafka_datasource)
+        elif message_format.lower() == "avro":
+            result = cudf.io.read_avro(self.kafka_datasource)
+        elif message_format.lower() == "parquet":
+            result = cudf.io.read_parquet(self.kafka_datasource)
```
Comment (kkraus14) [inline python/custreamz/custreamz/kafka.py:81]: Maybe have a dictionary that looks like: `{"json": cudf.io.read_json, "csv": cudf.io.read_csv, ...}` and use that instead of these `if` statements

File: python/custreamz/custreamz/kafka.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+        else:
+            raise ValueError(
+                "Unsupported Kafka Message payload type of: "
+                + str(message_format)
```
Comment (kkraus14) [inline python/custreamz/custreamz/kafka.py:85]: use f-string instead of string concatenation here

File: python/custreamz/custreamz/kafka.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        delimiter="\n",
+        message_format="json",
+    ):
+
+        """{docstring}"""
+
+        if topic is None:
+            raise ValueError(
+                "ERROR: You must specifiy the topic "
+                + "that you want to consume from"
```
Comment (kkraus14) [inline python/custreamz/custreamz/kafka.py:59]: Don't need the `+` here, can just do a multiline string

File: python/custreamz/custreamz/kafka.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+            result = cudf.io.read_parquet(self.kafka_datasource)
+        else:
+            raise ValueError(
+                "Unsupported Kafka Message payload type of: "
+                + str(message_format)
+            )
+
+        # Close up the cudf datasource instance
+        kafka_datasource.unsubscribe()
+        kafka_datasource.close()
```
Comment (kkraus14) [inline python/custreamz/custreamz/kafka.py:90]: Can we add a todo to handle these using some type of destructor / finalizer instead of relying on calling them explicitly?

File: python/custreamz/setup.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+import versioneer
+
+install_requires = ["cudf_kafka", "cudf"]
+
+CUDF_ROOT = os.environ.get("CUDF_ROOT", "../../cpp/build/")
+CUDF_KAFKA_ROOT = os.environ.get(
+    "CUDF_KAFKA_ROOT", "../../libcudf_kafka/build"
+)
+CUSTREAMZ_ROOT = os.environ.get("CUSTREAMZ_ROOT", "../../custreamz/build")
```
Comment (kkraus14) [inline python/custreamz/setup.py:14]: These all look unused?

File: python/custreamz/custreamz/kafka.py [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+        for part in partitions:
+            offset = self.kafka_meta_client.get_committed_offset(
+                part.topic, part.partition
+            )
+            if offset < 0:
+                offset = 0
+            toppars.append(
+                ck.TopicPartition(part.topic, part.partition, offset)
+            )
+        return toppars
```
Comment (kkraus14) [inline python/custreamz/custreamz/kafka.py:113]: Should refactor this into a list comprehension instead of creating an empty list and appending.

File: python/custreamz/custreamz/kafka.py [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+            )
+        except RuntimeError:
+            raise RuntimeError("Unable to connect to Kafka broker")
+
+        if len(offsets) != 2:
+            raise RuntimeError(
+                "Multiple watermark offsets encountered. "
+                + "Only 2 were expected and "
+                + str(len(offsets))
+                + " encountered"
```
Comment (kkraus14) [inline python/custreamz/custreamz/kafka.py:137]: Don't add strings with `+` and use f-string instead of `str(len(offsets))`

**Comment by kkraus14** [comment]
Also, if we want to actually build a `custreamz` package we'll need to update the conda recipe

**Comment by kkraus14** [comment]
There's also no unit testing of this code. Can we add some unit tests?

File: python/custreamz/custreamz/kafka.py (hunk 2 above)
Comment (jdye64) [reply python/custreamz/custreamz/kafka.py:21]: no particular reason. I will move them inline

**Comment by jdye64** [comment]
I agree on the unit tests part completely. However should we push that to another PR since we currently have no infrastructure for even the possibility of adding Kafka right now? I can make another issue for that very thing if you think that would be ok?

**Comment by kkraus14** [comment]
> I agree on the unit tests part completely. However should we push that to another PR since we currently have no infrastructure for even the possibility of adding Kafka right now? I can make another issue for that very thing if you think that would be ok?

It makes me really uncomfortable pushing all of this code without test coverage because users can't even test locally. Could we add some tests that are predicated on successfully connecting to a kafka cluster? I.E. in the test setup it connects to the kafka cluster, and if it fails to do so then it skips the entire testing module?

File: conda/recipes/custreamz/build.sh [hunk 14]
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.
+
+# This assumes the script is executed from the root of the repo directory
+
+# show environment
+printenv
+# Cleanup local git
+git clean -xdf
+# build libcudf with verbose output
```
Comment (kkraus14) [inline conda/recipes/custreamz/build.sh:9]:
```suggestion
# build custreamz with verbose output
```

File: python/custreamz/custreamz/kafka.py [hunk 15]
```diff
@@ -4,41 +4,84 @@
...
+        This object creates the underlying KafkaDatasource connection which
+        is used to read data from Kafka and create cudf Dataframes.
+        This class should not be directly instantiated.
+
+        Parameters
+        ----------
+        kafka_configs, dict, Key/Value pairs of librdkafka
+        configuration values. Full list of valid configuration
+        options can be found at
+        https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md
```
Comment (kkraus14) [inline python/custreamz/custreamz/kafka.py:23]: Need to correct formatting of docstrings, here's one for example:
```suggestion
        kafka_configs : dict
            Dict of key/value pairs of librdkafka configuration values. Full
            list of valid configuration options can be found at
            https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md
```
first line is argument name, type, and whether it's optional or not
second+ lines are indented and are a description of the parameter

File: python/custreamz/custreamz/kafka.py [hunk 16]
```diff
@@ -4,41 +4,84 @@
...
+
```
Comment (kkraus14) [inline python/custreamz/custreamz/kafka.py:24]:
```suggestion
```

File: python/custreamz/custreamz/kafka.py [hunk 17]
```diff
@@ -4,41 +4,84 @@
...
 
-        """{docstring}"""
+        """
+        Stop all active consumption and remove consumer subscriptions
+        to topic/partition instances
+
+        Parameters
+        ----------
+        None
+
```
Comment (kkraus14) [inline python/custreamz/custreamz/kafka.py:45]: If there's no parameters, you can omit the entire section
```suggestion
```

File: python/custreamz/custreamz/kafka.py [hunk 18]
```diff
@@ -95,27 +174,60 @@ def read_gdf(
...
+        Examples
+        --------
+        >>> from custream import kafka
+        >>> kafka_configs = {
+            "metadata.broker.list": "localhost:9092",
+            "enable.partition.eof": "true",
+            "group.id": "groupid",
+            "auto.offset.reset": "earliest",
+            "enable.auto.commit": "false"
+        }
```
Comment (kkraus14) [inline python/custreamz/custreamz/kafka.py:226]: I believe you need a `...` to have multiline code

**Comment by chinmaychandak** [comment]
`streamz` has some Kafka tests here: https://github.com/python-streamz/streamz/blob/master/streamz/tests/test_kafka.py. It uses the Spotify Kafka container.

**Comment by kkraus14** [comment]
> `streamz` has some Kafka tests here: https://github.com/python-streamz/streamz/blob/master/streamz/tests/test_kafka.py. It uses the Spotify Kafka container. Maybe we can add tests using the same container here?

We can't just add an arbitrary container to our CI system.

**Comment by chinmaychandak** [comment]
> We can't just add an arbitrary container to our CI system.

Ok. I was only suggesting a possible option.

**Comment by satishvarmadandu** [comment]
Can we use this - https://github.com/salesforce/kafka-junit ?

**Comment by harrism** [comment]
@jdye64 can you provide a more descriptive title (and description)?

File: ci/gpu/build.sh [hunk 19]
```diff
@@ -66,7 +66,8 @@ fi
 
 conda install "rmm=$MINOR_VERSION.*" "cudatoolkit=$CUDA_REL" \
               "rapids-build-env=$MINOR_VERSION.*" \
-              "rapids-notebook-env=$MINOR_VERSION.*"
+              "rapids-notebook-env=$MINOR_VERSION.*" \
+              "python-confluent-kafka"
```
Comment (kkraus14) [inline ci/gpu/build.sh:70]: Needs a PR into https://github.com/rapidsai/integration as this won't be merged until this line is removed

File: python/custreamz/custreamz/kafka.py [hunk 20]
```diff
@@ -17,7 +17,7 @@ def __init__(self, kafka_configs):
 
         Parameters
         ----------
-        kafka_configs, dict, Key/Value pairs of librdkafka
+        kafka_configs : dict, Dict of Key/Value pairs of librdkafka
         configuration values. Full list of valid configuration
         options can be found at
         https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md
```
Comment (kkraus14) [inline python/custreamz/custreamz/kafka.py]: This is still misformatted, first line is parameter name and type, second+ lines are indented and description of parameter:
```suggestion
        kafka_configs : dict,
            Dict of Key/Value pairs of librdkafka configuration values. Full
            list of valid configuration options can be found at
            https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md
```

Comment (kkraus14) [reply python/custreamz/custreamz/kafka.py]: This needs to be fixed for all the docstrings

File: python/custreamz/custreamz/tests/test_kafka.py [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+@pytest.fixture(scope="module")
+def client():
+    kafka_configs = {
+        "metadata.broker.list": "localhost:9092",
+        "enable.partition.eof": "true",
+        "group.id": "groupid",
+        "auto.offset.reset": "earliest",
+        "enable.auto.commit": "false",
+    }
+    return kafka.Consumer(kafka_configs)
```
Comment (kkraus14) [inline python/custreamz/custreamz/tests/test_kafka.py:19]: Move this into a `conftest.py` similar to https://github.com/rapidsai/cudf/blob/branch-0.15/python/cudf/cudf/tests/conftest.py and wrap it in a try except, where if it fails to create the client because there's no kafka cluster running then it does a `pytest.skip` instead. This way every test will be automatically skipped if there's no kafka cluster.

File: python/custreamz/custreamz/kafka.py (hunk 20 above)
Comment (kkraus14) [reply python/custreamz/custreamz/kafka.py]: Additionally, there's still docstrings with `Parameters None` and empty line breaks that should be fixed as well

File: python/custreamz/custreamz/tests/conftest.py [hunk 22]
```diff
@@ -1,22 +1,29 @@
...
+        if result == 0:
+            print("Kafka service is running")
+        else:
+            print("Kafka service is NOT running")
+        s.shutdown(2)
+        s.close()
+    except Exception:
+        pytest.skip(
+            "A running Kafka instance must be" " available to run these tests"
+        )
```
Comment (kkraus14) [inline python/custreamz/custreamz/tests/conftest.py:25]: Instead of doing this what if we just used a try except surrounding the `kafka.Consumer(kafka_configs)` call? I.E.
```
try:
    consumer = kafka.Consumer(kafka_configs)
except ConnectionError:  #This is a made up exception type
    pytest.skip("...")

return consumer
```

Comment (jdye64) [reply python/custreamz/custreamz/tests/conftest.py:25]: This proves difficult because if a Kafka instance is not running the call to attempt to create the Kafka consumer will infinitely attempt to reconnect, therefore never throwing an exception or returning. It does this because with broker rebalancing it allows for servers to be taken down or reassigned so it assumes you always provide the appropriate connection details and the server just isn't back online yet. 

It is possible to add an "event_cb" and try to parse that information out and throw an exception but that seems really error prone since indeed the lose of the connection could be something valid like a busy server not responding but a cluster being reconfigured.

Basically checking by socket is the only guaranteed way I can think of to ensure its just "down" and not something else expected going on.

File: python/custreamz/custreamz/kafka.py [hunk 23]
```diff
@@ -0,0 +1,280 @@
...
+        partitions : list,
+            Topic/Partition instances that specify the TOPPAR
+            instances the offsets should be retrieved for
+        timeout : int, default 10000,
+            Max time to wait on the response from
+            the Kafka broker in milliseconds
+
+        Returns
+        -------
+        Tuple of ck.TopicPartition objects
```
Comment (kkraus14) [inline python/custreamz/custreamz/kafka.py:194]:
```suggestion
        tuple
            Tuple of ck.TopicPartition objects
```

File: python/custreamz/custreamz/tests/conftest.py (hunk 22 above)
Comment (kkraus14) [reply python/custreamz/custreamz/tests/conftest.py:25]: Sounds good to me

File: python/custreamz/custreamz/tests/conftest.py [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+        result = s.connect_ex(("localhost", 9092))
+        if result == 0:
+            print("Kafka service is running")
+        else:
+            print("Kafka service is NOT running")
+        s.shutdown(2)
+        s.close()
+    except Exception:
+        pytest.skip(
+            "A running Kafka instance must be" " available to run these tests"
```
Comment (kkraus14) [inline python/custreamz/custreamz/tests/conftest.py:24]:
```suggestion
            "A running Kafka instance must be available to run these tests"
```

File: python/custreamz/custreamz/tests/conftest.py [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+def kafka_client():
+
+    # Check for the existance of a kafka broker
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    try:
+        result = s.connect_ex(("localhost", 9092))
+        if result == 0:
+            print("Kafka service is running")
+        else:
+            print("Kafka service is NOT running")
```
Comment (kkraus14) [inline python/custreamz/custreamz/tests/conftest.py:19]: Remove these print statements?
```suggestion
```

File: python/custreamz/custreamz/kafka.py [hunk 26]
```diff
@@ -0,0 +1,280 @@
...
+
```
[the remark this answers is not in the document]
Comment (jdye64) [reply python/custreamz/custreamz/kafka.py:195]: sorry I thought those spaces were desired. I will remove them all.

File: python/custreamz/custreamz/kafka.py [hunk 27]
```diff
@@ -0,0 +1,280 @@
...
+
+        topic=None,
+                partition=0,
+                lines=True,
+                start=0,
+                end=0,
+                batch_timeout=10000,
+                delimiter="\n",
+                message_format="json",
```
[the remark this answers is not in the document]
Comment (jdye64) [reply python/custreamz/custreamz/kafka.py:103]: looks like I accidentally pasted that there when moving around some configuration dicts. Removing it.

**Comment by jdye64** [comment]
reruns tests

**Comment by jdye64** [comment]
rerun tests

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 28]
```diff
@@ -112,6 +133,111 @@ void kafka_consumer::consume_to_buffer()
...
+
+  int64_t offset = toppar_list[0]->offset();
+  if (offset > 0) {
+    return offset;
+  } else {
+    return 0;
+  }
+}
+
+std::map<std::string, int64_t> kafka_consumer::get_watermark_offset(std::string topic,
```
Comment (davidwendt) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:169]:
```suggestion
std::map<std::string, int64_t> kafka_consumer::get_watermark_offset(std::string const& topic,
```

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 29]
```diff
@@ -112,6 +133,111 @@ void kafka_consumer::consume_to_buffer()
...
+    err = consumer->query_watermark_offsets(topic, partition, &low, &high, timeout);
+  }
+
+  if (err != RdKafka::ErrorCode::ERR_NO_ERROR) {
+    printf("Error: '%s'\n", err2str(err).c_str());
+    if (err == RdKafka::ErrorCode::ERR__PARTITION_EOF) {
+      results.insert(std::pair<std::string, int64_t>("low", low));
+      results.insert(std::pair<std::string, int64_t>("high", high));
+    } else {
+      throw std::runtime_error(std::string(err2str(err).c_str()));
```
Comment (davidwendt) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:191]: Seems like we are creating alot of temporary strings here.
Should use `CUDF_FAIL` here instead?

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 30]
```diff
@@ -112,6 +133,111 @@ void kafka_consumer::consume_to_buffer()
...
+    }
+  } else {
+    results.insert(std::pair<std::string, int64_t>("low", low));
+    results.insert(std::pair<std::string, int64_t>("high", high));
+  }
+
+  return results;
+}
+
+bool kafka_consumer::commit_offset(std::string topic, int partition, int64_t offset)
```
Comment (davidwendt) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:201]:
```suggestion
bool kafka_consumer::commit_offset(std::string const& topic, int partition, int64_t offset)
```
Requires changing the declaration in the header as well.

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 31]
```diff
@@ -24,6 +24,27 @@ namespace io {
 namespace external {
 namespace kafka {
 
+kafka_consumer::kafka_consumer(std::map<std::string, std::string> configs)
```
Comment (davidwendt) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:27]:
```suggestion
kafka_consumer::kafka_consumer(std::map<std::string, std::string> const& configs)
```

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 32]
```diff
@@ -112,6 +133,110 @@ void kafka_consumer::consume_to_buffer()
...
+    key = (*it);
+    it++;
+    val = (*it);
+    it++;
+    configs.insert(std::pair<std::string, std::string>{key, val});
+  }
+  return configs;
+}
+
+int64_t kafka_consumer::get_committed_offset(std::string topic, int partition)
```
Comment (davidwendt) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:152]:
```suggestion
int64_t kafka_consumer::get_committed_offset(std::string const& topic, int partition)
```

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 33]
```diff
@@ -112,6 +133,110 @@ void kafka_consumer::consume_to_buffer()
...
+{
+  std::vector<RdKafka::TopicPartition *> toppar_list;
+  // toppar_list.push_back(find_toppar(topic, partition));
```
Comment (davidwendt) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:155]: Remove commented out code

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 34]
```diff
@@ -112,6 +133,110 @@ void kafka_consumer::consume_to_buffer()
...
+{
+  RdKafka::ErrorCode err = consumer.get()->close();
+  if (err != RdKafka::ERR_NO_ERROR) {
+    printf("Timeout occurred while closing Kafka Consumer\n");
+    return false;
+  } else {
+    return true;
+  }
+  delete consumer.get();
+  delete kafka_conf.get();
```
Comment (davidwendt) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:237]: These two lines will never execute. I would've expected a compiler error/warning.

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 35]
```diff
@@ -112,6 +133,110 @@ void kafka_consumer::consume_to_buffer()
...
+
+  // Query Kafka to populate the TopicPartitions with the desired offsets
+  RdKafka::ErrorCode err = consumer->committed(toppar_list, default_timeout);
+
+  int64_t offset = toppar_list[0]->offset();
+  if (offset > 0) {
+    return offset;
+  } else {
+    return 0;
+  }
```
Comment (davidwendt) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:166]: Use ternary here:
```
   return offset > 0 ? offset : 0;
```

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 36]
```diff
@@ -112,6 +133,110 @@ void kafka_consumer::consume_to_buffer()
...
+
+bool kafka_consumer::close(int timeout)
+{
+  RdKafka::ErrorCode err = consumer.get()->close();
+  if (err != RdKafka::ERR_NO_ERROR) {
+    printf("Timeout occurred while closing Kafka Consumer\n");
+    return false;
+  } else {
+    return true;
+  }
```
Comment (davidwendt) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:235]: I like to use the pattern as such:
```
  if (err != RdKafka::ERR_NO_ERROR) {
    printf("Timeout occurred while closing Kafka Consumer\n");
    return false;
  }
  return true;
```
There is no need for the `else` and it would've made the abandoned remaining lines more obvious.

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 37]
```diff
@@ -24,7 +24,28 @@ namespace io {
...
+  CUDF_EXPECTS(RdKafka::Conf::ConfResult::CONF_OK == kafka_conf->get("group.id", conf_val),
+               "Kafka group.id must be configured");
+
+  std::string errstr;
+  consumer = std::unique_ptr<RdKafka::KafkaConsumer>(
+    RdKafka::KafkaConsumer::create(kafka_conf.get(), errstr));
+}
+
+kafka_consumer::kafka_consumer(std::map<std::string, std::string> const &configs,
                                std::string topic_name,
```
Comment (davidwendt) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:49]:
```suggestion
                               std::string const& topic_name,
```

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 38]
```diff
@@ -112,6 +133,102 @@ void kafka_consumer::consume_to_buffer()
...
+}
+
+bool kafka_consumer::close(int timeout)
+{
+  RdKafka::ErrorCode err = consumer.get()->close();
+  delete consumer.get();
+  delete kafka_conf.get();
+
+  if (err != RdKafka::ERR_NO_ERROR) { return false; }
+  return true;
```
Comment (davidwendt) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:229]: How about this instead?
```
  return err == RdKafka::ERR_NO_ERROR;
```

File: cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp [hunk 39]
```diff
@@ -97,6 +168,7 @@ class kafka_consumer : public cudf::io::datasource {
   int64_t start_offset;
   int64_t end_offset;
   int batch_timeout;
+  int default_timeout = 10000;
```
Comment (davidwendt) [inline cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp:171]:
```suggestion
  int default_timeout = 10000; // milliseconds
```

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 40]
```diff
@@ -112,6 +133,101 @@ void kafka_consumer::consume_to_buffer()
...
+  std::string val;
+  for (std::list<std::string>::iterator it = dump->begin(); it != dump->end();) {
+    key = (*it);
+    it++;
+    val = (*it);
+    it++;
+    configs.insert(std::pair<std::string, std::string>{key, val});
+  }
+  return configs;
+}
```
Comment (harrism) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:150]: This can be simplified a lot. @codereport and I brainstormed this together
https://godbolt.org/z/TMc39P

```suggestion
std::map<std::string, std::string> kafka_consumer::current_configs()
{
  std::map<std::string, std::string> configs;
  std::list<std::string> *dump = kafka_conf->dump();
  for (auto it = dump->begin(); it != dump->end(); std::advance(it, 2))
    configs.insert({*it, *std::next(it)});
  return configs;
}
```

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 41]
```diff
@@ -112,6 +133,101 @@ void kafka_consumer::consume_to_buffer()
...
+    return false;
+  } else {
+    return true;
+  }
+}
+
+bool kafka_consumer::close(int timeout)
+{
+  RdKafka::ErrorCode err = consumer.get()->close();
+  delete consumer.get();
```
[the remark this answers is not in the document]
Comment (jdye64) [reply cpp/libcudf_kafka/src/kafka_consumer.cpp:225]: @harrism Just want to make sure I'm not missing something but this should be `consumer.reset(nullptr)` and not `consumer->reset(nullptr)` which would attempt to dereference and call the function `reset` on the `consumer` object right?

[the remark this answers is not in the document]
Comment (jdye64) [reply cpp/libcudf_kafka/src/kafka_consumer.cpp:225]: also since both `consumer` and `kafka_conf` are C implementations would I not need to explicitly call `delete` on them? This is from the Kafka docs [here](https://docs.confluent.io/2.0.0/clients/librdkafka/classRdKafka_1_1KafkaConsumer.html#a5c78a721aa91f3be9903f09ddf084644) at the bottom of the page, sorry can get an embedded link for it but it says ... `The consumer object must later be freed with delete`

**Comment by harrism** [comment]
C++ looks great now!

File: cpp/libcudf_kafka/src/kafka_consumer.cpp (hunk 41 above)
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/libcudf_kafka/src/kafka_consumer.cpp]: https://en.cppreference.com/w/cpp/memory/unique_ptr/reset

> 1) Given current_ptr, the pointer that was managed by *this, performs the following actions, in this order:
Saves a copy of the current pointer old_ptr = current_ptr
Overwrites the current pointer with the argument current_ptr = ptr
If the old pointer was non-empty, deletes the previously managed object if(old_ptr) get_deleter()(old_ptr).

Now, `get_deleter()` is going to return `std::default_delete<T>` (https://en.cppreference.com/w/cpp/memory/default_delete)

Which says: 
> 1) The non-specialized default_delete uses delete to deallocate memory for a single object.

So this means it will do the equivalent of the code you had before, except that now the `unique_ptr` will not have a dangling pointer to the thing you deleted.

File: python/cudf_kafka/cudf_kafka/_lib/kafka.pxd [hunk 42]
```diff
@@ -33,3 +55,10 @@ cdef class KafkaDatasource(Datasource):
     cdef string delimiter
 
     cdef datasource* get_datasource(self) nogil
+
+    cpdef bool commit_offset(self,
+                             string topic,
+                             int32_t partition,
+                             int64_t offset)
```
Comment (kkraus14) [inline python/cudf_kafka/cudf_kafka/_lib/kafka.pxd:62]: Should this be made `void` as well to align with the C++ class?

Comment (jdye64) [reply python/cudf_kafka/cudf_kafka/_lib/kafka.pxd:62]: yes, I build locally and it didn't fail which has me confused though? I do agree it needs to be changed though

File: python/cudf_kafka/cudf_kafka/_lib/kafka.pyx [hunk 43]
```diff
@@ -11,20 +12,37 @@ cdef class KafkaDatasource(Datasource):
...
 
     cdef datasource* get_datasource(self) nogil:
         return <datasource *> self.c_datasource.get()
+
+    cpdef bool commit_offset(self,
+                             string topic,
+                             int32_t partition,
+                             int64_t offset):
+        (<kafka_consumer *> self.c_datasource.get()).commit_offset(
+            topic, partition, offset)
```
Comment (kkraus14) [inline python/cudf_kafka/cudf_kafka/_lib/kafka.pyx:42]:
```suggestion
    cpdef void commit_offset(self,
                             string topic,
                             int32_t partition,
                             int64_t offset):
        (<kafka_consumer *> self.c_datasource.get()).commit_offset(
            topic, partition, offset)
```
